### ISSUE TITLE
🐛 correctness/noUnusedVariables inferring in unions reports unused variables

### ISSUE THREAD

**Issue by Louis-Michalak** [biomejs/biome#565]
### Environment information

```block
CLI:
  Version:                      1.3.0
  Color support:                true

Platform:
  CPU Architecture:             aarch64
  OS:                           macos

Environment:
  BIOME_LOG_DIR:                unset
  NO_COLOR:                     unset
  TERM:                         "xterm-256color"
  JS_RUNTIME_VERSION:           "v18.18.2"
  JS_RUNTIME_NAME:              "node"
  NODE_PACKAGE_MANAGER:         "yarn/1.22.19"

Biome Configuration:
  Status:                       Loaded successfully
  Formatter disabled:           false
  Linter disabled:              false
  Organize imports disabled:    false
  VCS disabled:                 true

Workspace:
  Open Documents:               0
```

### What happened?

Infering in unions reports unused variables on leading infered type variables.

```typescript
// Nonesense type, just so you get the idea
type TestUnionType<T> = T extends (infer B)[] | infer B ? B : never;
//                                       ^
//                                       "This variable is unused"

type Value = TestUnionType<"">; // ""
type Value2 = TestUnionType<["", "okok"]>; // "" | "okok"
```

### Expected result

noUnusedVariables shouldn't be reported

### Code of Conduct

- [X] I agree to follow Biome's Code of Conduct

**Comment by Conaclos** [comment]
This is a tricky case. TypeScript has so many exceptions...

Our name resolver has the assumption that a binding is bound to a single declaration. Here `B` is declared at different places. I am unsure how to fix the issue. We could ignore extra declarations or turn them into reads, however this conflicts with another assumption: every binding identifier has an associated semantic binding.

@denbezrukov @ematipico any thoughts?

**Comment by ematipico** [comment]
I might not know few things, so I might get things wrong. 

Given this code 

```ts
T extends (infer B)[] | infer B ? B : never
```

Isn't `B` declared twice in two different scopes? On in `(infer B)[]` and the second in `infer B ? B : never`?

**Comment by Conaclos** [comment]
Unfortunately not, it is the same declaration :/

If you hover and click on the B reference in [TypeScript Playground](https://www.typescriptlang.org/play?#code/C4TwDgpgBAKhDOwCqA7AlgexTcEA8MAfFALyxQQAewEKAJvFABRooBmEATlAEICUAbQC6UAD5RWHbjwBQUeVAD8vOQoBcUFBABuXANxA), you notice that the reference is bound to these two declarations.

**Comment by Conaclos** [comment]
On reflection, we could bind multiple binding identifiers to the same semantic binding. This could solve our problem without destroying all our assumptions. This will require a refactoring of how we identify a semantic binding (for now we identify it using the binding identifier's scope). I am not 100% happy about this because it prevents some optimisations I was planning to do. However, this seems like the best compromise to me.

**Comment by Conaclos** [comment]
Instead of introducing exceptions in the semantic model, we could just ignore `infer T` that precedes another `infer T`.

**Comment by ematipico** [comment]
Yeah, we could try this approach
